dev/src/Command/AddComponentCommand.php

```diff
@@ -184,21 +210,215 @@ protected function execute(InputInterface $input, OutputInterface $output)
+        } else if ($input->getOption('googleapis-gen-path')) {
+            $this->checkDockerAvailable();
+            $googleApisGenDir = realpath($input->getOption('googleapis-gen-path'));
+            $output->writeln("\n\nCopying the library code from googleapis-gen");
+            $output->writeln($this->owlbotCopyCode($new->componentName, $googleApisGenDir));
+            $output->writeln("\n\nOwlbot post processing");
+            $output->writeln($this->owlbotPostProcessor());
+        }
+
         $output->writeln('');
         $output->writeln('');
         $output->writeln('Success!');
 
         return 0;
     }
 
+    private function validateOptions(InputInterface $input, OutputInterface $output): void
+    {
+        $error = '<error>googleapis-gen-path and bazel-path options ' .
+            'are mutually exclusive. Will build library with bazel.</error>';
+        if ($input->getOption('bazel-path') && $input->getOption('googleapis-gen-path')) {
+            $output->writeln('');
+            $output->writeln($error);
+            $output->writeln('');
+        }
+    }
+
+    private function loadYamlConfigContent(NewComponent $new, string $protoDir): ?array
+    {
+        $yamlFilePath = sprintf('%s/%s%s%s.yaml',
+            $protoDir,
+            strtolower($new->shortName),
+            $new->version ? '_' : '',
+            $new->version ?? ''
+        );
+
+        try {
+            return Yaml::parse($this->loadProtoContent($yamlFilePath));
+        } catch (Exception $e) {
+            // Handle error gracefully.
+            return null;
+        }
+    }
+
     private function loadProtoContent(string $proto): string
     {
         if (file_exists($proto)) {
             return file_get_contents($proto);
         }
         $protoUrl = 'https://raw.githubusercontent.com/googleapis/googleapis/master/' . $proto;
-        $client = new Client();
-        $response = $client->get($protoUrl);
+        $response = $this->httpClient->get($protoUrl);
         return (string) $response->getBody();
     }
+
+    private function bazelQueryAndBuildLibrary(string $protoPath, string $googleApisDir): string
+    {
+        $command = ['bazel', '--version'];
+        $output = $this->runCommand($command);
+        // Extract the version number from the output
+        $match = preg_match('/bazel\s+(?P<version>\d+\.\d+\.\d+)/', $output, $matches);
+        if (!$match || $matches['version'] !== self::BAZEL_VERSION) {
+            throw new RuntimeException('Bazel 6.0.0 is not available');
+        }
+
+        $componentPath = str_replace(['(', ')'], '', $protoPath);
+        $command = [
+            'bazel',
+            'query',
+            'filter("-(php)$", kind("rule", //' . $componentPath .'/...:*))'
+        ];
+        $output = $this->runCommand($command, $googleApisDir);
+        $command = ['grep', '-v', '-E', ':(proto|grpc|gapic)-.*-php$'];
+        $output = $this->runCommand($command, $googleApisDir, $output);
```

Comment:
nit: we're shelling out in order to use the `grep` command, which is a bit odd, since PHP has built-in functions that you could use instead. And because it seems like you're trying to sanitize bazel's output, it would make more sense to do something like this:

```php
$components = array_filter(
    explode("\n", $output),
    fn ($line) => !preg_match('/:(proto|grpc|gapic)-.*-php$/', $line)
);
```

... and if I understand this code correctly, we could then throw an exception if there is more than one line (because there shouldn't be, right?)

```php
if (count($components) > 1) {
    throw new Exception('more than one bazel component found: ' . implode(' ', $components));
} 

return $this->bazelBuildLibrary($components[0]);
```

Alternatively, you could chain the `grep` command to the `bazel` command, since you're already shelling out, but to shell out just for grep is unnecessary (because we try to only shell out for things that we absolutely must shell out for).

Reply:
Great, changed the following way to remove the empty lines as well:
```
$components = array_filter(
            explode("\n", $output),
            fn ($line) => !preg_match('/:(proto|grpc|gapic)-.*-php$/', $line) && !empty($line)
);
```

Reply:
Good to know about chaining with `Process`, I also wasn't sure if that was possible.

I think filtering in PHP works great though - if we were worried about performance, I would suggest figuring out how to pipe.

As for the empty line, I think that's great, although I'd suggest placing the check for `$line` _before_ the `preg_match`, since it's the faster check, e.g.

```php
fn ($line) => $line && !preg_match('/:(proto|grpc|gapic)-.*-php$/', $line)
```

Reply:
I figured out how to do pipes! Use the `Process::fromShellCommandline` function, which accepts a raw string as input. The downside is there won't be shell escaping on individual arguments, so you need to be careful and also use the `"${:MY_VAR}"` syntax to sanitize variables. It looks like this:

```php
$p = Process::fromShellCommandline('ls -la | grep php');
$process->run();
```

So in our case something like this might work:

```php
$command = 'bazel query filter("-(php)$", kind("rule", //"${:path}"/...:*)) '
    . '| grep -v -E ":(proto|grpc|gapic)-.*-php$"';
$process = Process::fromShellCommandline('ls -la | grep php');
$process->run(null, ['path' => $componentPath]);
```

Although I haven't tested it, and think the way we have works well, this might be something you prefer (if it works).

Reply:
I am good with the current implementation, will stick to that. However thank you for sharing this, this is good information.